core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2597,14 +2597,30 @@ private[spark] object Utils extends Logging {
   }
 
   /**
-   * Push based shuffle can only be enabled when the application is submitted
-   * to run in YARN mode, with external shuffle service enabled
+   * Push based shuffle can only be enabled when below conditions are met:
+   *   - the application is submitted to run in YARN mode
+   *   - external shuffle service enabled
+   *   - IO encryption disabled
+   *   - serializer(such as KryoSerializer) supports relocation of serialized objects
    */
   def isPushBasedShuffleEnabled(conf: SparkConf): Boolean = {
-    conf.get(PUSH_BASED_SHUFFLE_ENABLED) &&
-      (conf.get(IS_TESTING).getOrElse(false) ||
-        (conf.get(SHUFFLE_SERVICE_ENABLED) &&
-          conf.get(SparkLauncher.SPARK_MASTER, null) == "yarn"))
+    val serializer = Utils.classForName(conf.get(SERIALIZER)).getConstructor(classOf[SparkConf])
+      .newInstance(conf).asInstanceOf[Serializer]
+    val canDoPushBasedShuffle =
+      conf.get(PUSH_BASED_SHUFFLE_ENABLED) &&
+        (conf.get(IS_TESTING).getOrElse(false) ||
+          (conf.get(SHUFFLE_SERVICE_ENABLED) &&
+            conf.get(SparkLauncher.SPARK_MASTER, null) == "yarn" &&
+            // TODO: [SPARK-36744] needs to support IO encryption for push-based shuffle
+            !conf.get(IO_ENCRYPTION_ENABLED) &&
+            serializer.supportsRelocationOfSerializedObjects))
+
+    if (!canDoPushBasedShuffle) {
+      logWarning("Push-based shuffle can only be enabled when the application is submitted" +
```

Comment:
nit. We need a space at the end of the string.

Reply:
This has been fixed in the follow-up PR #33984.